Feat: Wishlist management

Rohitprasad83 opened the pull request:
- Added wishlist functionality using useState and useContext Hooks
- Improved styling of the wishlist page
- Not yet added move to wishlist in Cart page, will do it once cart management is completed.
- Gif :-

![chrome_OifoQ7rLGQ](https://user-images.githubusercontent.com/66221637/160234088-cb6d458b-571c-418a-ab02-2a15bb7adbbd.gif)

Rohitprasad83 committed:
Feat: added wishlist context

Rohitprasad83 committed:
feat: imported WishlistContextProvider to index.js

Rohitprasad83 committed:
feat: basic wishlist management completed

Rohitprasad83 committed:
feat: the Color of wishlist icon changes accordingly

Rohitprasad83 committed:
refactor: the wishlist is attached to heart icon

Rohitprasad83 committed:
feat: navbar wishlist badge count equal to wishlist items

Rohitprasad83 committed:
feat: add to wishlist from product page

Rohitprasad83 committed:
feat: wishlist page with wishlist products

Rohitprasad83 committed:
feat: navigation to products page when wishlist is empty

Rohitprasad83 committed:
style: improved styling for wishlist page

Rohitprasad83 committed:
Remove: unwanted comment

jeetbhalani15 reviewed (approved):
some suggestions u could have use useReducer too.
u can have one provider which contains all other providers.
nice ui

Naveenchandar commented on src/components/navbar/Navbar.jsx:
```diff
@@ -19,7 +21,10 @@ export function Navbar() {
               <i className="fa-regular fa-heart icon"> </i>
-              <span className="badge__number badge__primary"> 5 </span>
+              <span className="badge__number badge__primary">
+                {' '}
```
You can remove this empty curly braces. This is not needed.

Rohitprasad83 replied:
Prettier was putting them, now I have removed them

Naveenchandar commented on src/index.js:
```diff
@@ -4,18 +4,20 @@ import './index.css'
+      <WishlistProvider>
+        <ProductFilterProvider>
+          <App />
+        </ProductFilterProvider>{' '}
+      </WishlistProvider>{' '}
+    </BrowserRouter>{' '}
```
You can remove curly braces. This is not needed.

Rohitprasad83 replied:
Done

Naveenchandar commented on src/pages/products/ProductCard.jsx:
```diff
@@ -10,10 +10,41 @@ export function ProductCard({
+    rating,
+  }
+  function addWishlistItem() {
+    for (let i = 0; i < wishlistItems.length; i++) {
+      if (item._id === wishlistItems[i]._id)
+        return setWishlistItems(
+          wishlistItems.filter(product => item._id !== product._id)
+        )
+    }
```
Try to use es6 for in or for of loop or array.map.

Rohitprasad83 replied:
replaced for loop to for of loop

Naveenchandar commented on src/pages/wishlist/Wishlist.jsx:
```diff
@@ -11,11 +15,31 @@ export function Wishlist() {
-          <WishlistCard />
-          <WishlistCard />
+          {wishlistItems.map(product => (
+            <WishlistCard key={product['id']} {...product} />
```
any specific reason for not using dot operator for specifying object values ?

Rohitprasad83 replied:
No, but now using dot operator

Naveenchandar commented on src/pages/wishlist/WishlistCard.jsx:
```diff
@@ -1,22 +1,55 @@
+  const { _id, title, img, price, oldPrice, discount, categoryName, rating } =
+    product
+
+  function addWishlistItem() {
+    for (let i = 0; i < wishlistItems.length; i++) {
+      if (product._id === wishlistItems[i]._id)
+        return setWishlistItems(
+          wishlistItems.filter(({ _id }) => _id !== product._id)
+        )
+    }
+    setWishlistItems([...wishlistItems, product])
+  }
+  function containsInWishlist(_id) {
+    return wishlistItems.some(product => product._id === _id)
+  }
```
Please lift up this to wishlist context. It will avoid repetitive code.

Rohitprasad83 replied:
Thanks for suggesting, I have now put this functions into utils folder. Please review and approve the pull request

Rohitprasad83 committed:
Removed: unwanted {' '}

Rohitprasad83 committed:
feat: added common wishlist functions in utils folder

Rohitprasad83 committed:
refactor: used utils function for wishlist

using addWishlistItem and containsInWishlist util function in productCard and wishlistCard

Rohitprasad83 committed:
Refactor: using dot operator in ProductCard Component
